include/experimental/__p0009_bits/layout_left.hpp

```diff
@@ -116,18 +116,19 @@ class layout_left::mapping {
     )
     MDSPAN_CONDITIONAL_EXPLICIT((extents_type::rank() > 0))
     MDSPAN_INLINE_FUNCTION _MDSPAN_CONSTEXPR_14
-    mapping(layout_stride::mapping<OtherExtents> const& other) // NOLINT(google-explicit-constructor)
+    mapping(layout_stride::mapping<OtherExtents> const& other) noexcept // NOLINT(google-explicit-constructor)
       :__extents(other.extents())
     {
        /*
         * TODO: check precondition
         * other.required_span_size() is a representable value of type index_type
         */
-       #ifndef __CUDA_ARCH__
-       size_t stride = 1;
+       #if !defined(_MDSPAN_HAS_CUDA) && !defined(_MDSPAN_HAS_HIP) && !defined(NDEBUG)
+       index_type stride = 1;
        for(rank_type r=0; r<__extents.rank(); r++) {
-         if(stride != other.stride(r))
+         if(stride != static_cast<index_type>(other.stride(r)))
            throw std::runtime_error("Assigning layout_stride to layout_left with invalid strides.");
+           // Note this throw will lead to a terminate if triggered since this function is marked noexcept
```

Comment:
To clarify @dalg24 's comment: `assert` does something exactly when `not defined(NDEBUG)` is `true`, so this should behave effectively like `assert`.  Furthermore, an actual `assert` may have a better error message, as a `noexcept` violation might not need to print the exception message.

The nicer way to do this would be to define an `MDSPAN_ASSERT` macro that Does The Right Thing, but I think it's fine for now just to use `assert`.

Reply:
immediately invoked lambda code was worse.  reverted to the initial PR